src/mgr/ServiceMap.cc

```diff
@@ -75,20 +76,105 @@ std::string ServiceMap::Service::get_summary() const
   if (daemons.empty()) {
     return "no daemons active";
   }
+
   std::ostringstream ss;
-  ss << daemons.size() << (daemons.size() > 1 ? " daemons" : " daemon")
-     << " active";
+  /*
+   * The format is "type/prefix:daemon_name".
+   * "type": will be used to replace the default "daemon(s)"
+   * showed in `ceph -s`. If absent, the "daemon" will used.
+   * "prefix": if present the active members will be classified
+   * by the "prefix" instead of "daemon_name".
+   *
+   * For exmaple for iscsi gateways, it will be:
+   * "portal/hostX:rbd/imageX"
+   * The `ceph -s` will be:
+   *     igw: 3 portals active (hostX, hostY, hostZ)
+   */
+  std::set<std::string> prefs;
+  bool classified = false;
+  auto p = daemons.begin();
+  std::regex pattern("(\\w*)/(\\w*):(\\S+)"); // full pattern
+  std::smatch match;
+  std::string type;
+  if (std::regex_search(p->first, match, pattern)) {
+    type = std::string(match[1]);
+    std::string prefix = std::string(match[2]);
+    std::string name = std::string(match[3]);
+    if (name.empty()) {
+      /* Invalid pattern, just show what they are one by one */
+    } else {
+      classified = true;
+
+      /* If the prefix is absent, will show the daemon name instead */
+      if (!prefix.empty())
+        prefs.insert(prefix);
+      else
+        prefs.insert(name);
+
+      while (++p != daemons.end()) {
+        if (!std::regex_search(p->first, match, pattern)) {
+          /*
+           * The key patterns are inconsistend for all the daemons,
+           * stop classifying the daemons
+           */
+          classified = false;
+          break;
+        }
+
+        if (type != std::string(match[1])) {
+          /*
+           * The types are inconsistend for all the daemons,
+           * stop classifying the daemons
+           */
+          classified = false;
+          break;
+	}
+
+	std::string _prefix = std::string(match[2]);
+        std::string _name = std::string(match[3]);
+
+	if (!!prefix.empty() != !!_prefix.empty() || _name.empty()) {
+          /*
+           * The key patterns are inconsistend for all the daemons,
+           * stop classifying the daemons
+           */
+          classified = false;
+          break;
+	}
 
-  if (!daemons.empty()) {
-    ss << " (";
+        /* If the prefix is absent, will show the daemon name instead */
+        if (!_prefix.empty())
+          prefs.insert(_prefix);
+        else
+          prefs.insert(_name);
```

Comment:
i'd suggest add curly brace around branch blocks even if there is only a single statement in them for better readability.

Reply:
Okay. np.